### PR TITLE
Search PATHEXT instead of 3 hardcoded values

### PR DESCRIPTION
**Pull request by isiahmeadows**
Redo of #134.

Since it was only a few lines, it was easy to rewrite against master.

### CODE REVIEW

File: src/which.js [hunk 1]
```diff
@@ -26,8 +29,9 @@ function checkPath(path) {
 //@ var nodeExec = which('node');
 //@ ```
 //@
-//@ Searches for `command` in the system's PATH. On Windows looks for `.exe`, `.cmd`, and `.bat` extensions.
-//@ Returns string containing the absolute path to the command.
```
Comment (nfischer) [inline src/which.js]: Why was this line removed?

**Comment by nfischer** [comment]
@isiahmeadows could you `generate-docs`?

File: src/which.js [hunk 2]
```diff
@@ -2,6 +2,9 @@ var common = require('./common');
 var fs = require('fs');
 var path = require('path');
 
+// XP's system default value for PATHEXT system variable
```
Comment (BYK) [inline src/which.js]: Might be good to mentioned that WinXP lacks this thing in the comment. Also not sure if supporting Windows XP is a good idea. It is quite old and unsupported even by Microsoft.

File: src/which.js [hunk 3]
```diff
@@ -47,20 +51,16 @@ function _which(options, cmd) {
...
-          where = attempt;
-          return;
-        }
-        attempt = baseAttempt + '.cmd';
-        if (checkPath(attempt)) {
-          where = attempt;
-          return;
+        var pathExtEnv = process.env.PATHEXT || XP_DEFAULT_PATHEXT;
+        var pathExtArray = splitPath(pathExtEnv);
+        var len = pathExtArray.length;
```
Comment (BYK) [inline src/which.js]: Don't think array length caching makes any difference so you can just use `pathExtArray.length` in the `for` part without any performance penalty.

**Comment by BYK** [comment]
LGTM

FYI: Windows also has a built-in `where` command.

File: src/which.js (hunk 2 above)
Comment (nfischer) [inline src/which.js]: I don't see much harm in the default value. I agree that WinXP doesn't need to be supported by us, since the OS is end-of-life by Microsoft. However, just in case it's perhaps possible to delete environmental variables in Windows (is this even possible?), It'd be good to keep some fall-back value.

File: src/which.js (hunk 3 above)
Comment (nfischer) [inline src/which.js]: I don't know if it's part of the standard or not. Either way is fine with me, since the meaning is very clear (and performance isn't super critical yet).

File: src/which.js (hunk 2 above)
Comment (BYK) [inline src/which.js]: You can delete environmental variables but if you delete this one I'm guessing shelljs would be the least of your problems :D

Comment (BYK) [inline src/which.js]: Still, a default value is a good fallback so we should keep it.

File: src/which.js (hunk 3 above)
Comment (BYK) [inline src/which.js]: There are posts over the Internet saying it is simply unnecessary unless you're dealing with very old IE versions. Both ways are fine.

File: src/which.js (hunk 2 above)
Comment (nfischer) [inline src/which.js]: ^ Good point. But yes, I also vote we keep it for no other reason than extra safety measure.

File: src/which.js [hunk 4]
```diff
@@ -26,8 +29,9 @@ function checkPath(path) {
 //@ var nodeExec = which('node');
 //@ ```
 //@
-//@ Searches for `command` in the system's PATH. On Windows looks for `.exe`, `.cmd`, and `.bat` extensions.
-//@ Returns string containing the absolute path to the command.
+//@ Searches for `command` in the system's PATH. On Windows, looks for extensions in the PATHEXT
```
Comment (nfischer) [inline src/which.js]: *Totally optional suggestion*

Perhaps we don't need this much info in the README? The README is supposed to give a brief overview of  commands, and this might be too much detail. I think it's sufficient to say something to the effect of *"On Windows, this uses the `PATHEXT` environment variable to append the extension"* here.

I think it's better to have this comment down below in the function body.

Comment (isiahmeadows) [inline src/which.js]: Will fix.

File: src/which.js (hunk 1 above)
Comment (isiahmeadows) [inline src/which.js]: By accident.

File: src/which.js (hunk 2 above)
Comment (isiahmeadows) [inline src/which.js]: I'm clarifying the comment as it's a precautionary measure.

File: src/which.js (hunk 3 above)
Comment (isiahmeadows) [inline src/which.js]: Nowadays, I normally don't cache the length unless I'm writing multiple loops of the same array in a performance sensitive section of code. This was originally a simple copy-paste of a 1 1/2 year old PR, where I was still in the habit.

(Cheap checks first are often faster in my experience unless you're dealing with large arrays. But that doesn't apply here.)

**Comment by isiahmeadows** [comment]
Do note that I use Linux, my Windows installation is rather broken ATM, and my computer is too old/underpowered to run a VM, so I don't know how well this runs, and I kinda feel guilty not being able to test this. :frowning: 

But feel free to test this yourself. I did add a test. It's also a trivial enough patch that it *should* work.

(@nfischer If you all got Appveyor to run for PRs as well, that'd be awesome, probably even for you, considering this supports Windows. :wink:)

I submitted the original PR, #134, about 1 1/2 years ago, when my Windows installation actually worked (mine needs *way* more troubleshooting than I have time for ATM - my standard user directory requires Admin privileges to even *read*).

**Comment by nfischer** [comment]
I'll test this out on my Windows machine today and see how it runs. Thanks again!

File: test/which.js [hunk 5]
```diff
@@ -23,8 +23,16 @@ assert.equal(result, null);
...
 assert.equal(shell.error(), null);
-assert.equal(fs.existsSync(result), true);
+assert.equal(fs.existsSync(node), true);
+
+if (process.platform === 'win32') {
+    // This should be equivalent on Windows
+    var nodeExe = shell.which('node.exe');
+    assert.equal(shell.error(), null);
+    assert.equal(fs.existsSync(nodeExe), true);
+    assert.equal(node, node.exe);
```
Comment (nfischer) [inline test/which.js]: typeo: s/node.exe/nodeExe

File: test/which.js [hunk 6]
```diff
@@ -23,8 +23,16 @@ assert.equal(result, null);
...
+var node = shell.which('node');
 assert.equal(shell.error(), null);
-assert.equal(fs.existsSync(result), true);
+assert.equal(fs.existsSync(node), true);
+
+if (process.platform === 'win32') {
+    // This should be equivalent on Windows
+    var nodeExe = shell.which('node.exe');
+    assert.equal(shell.error(), null);
+    assert.equal(fs.existsSync(nodeExe), true);
```
Comment (nfischer) [inline test/which.js]: nit: change to `assert.ok(fs.existsSync(node))`

File: test/which.js [hunk 7]
```diff
@@ -23,8 +23,16 @@ assert.equal(result, null);
 // Valids
 //
 
-var result = shell.which('node');
+var node = shell.which('node');
 assert.equal(shell.error(), null);
-assert.equal(fs.existsSync(result), true);
+assert.equal(fs.existsSync(node), true);
```
Comment (nfischer) [inline test/which.js]: nit: change to `assert.ok(fs.existsSync(node))`

File: src/which.js [hunk 8]
```diff
@@ -46,23 +51,32 @@ function _which(options, cmd) {
...
+        // In case the PATHEXT variable is somehow not set (e.g.
+        // child_process.spawn with an empty environment), use the XP default.
+        var pathExtEnv = process.env.PATHEXT || XP_DEFAULT_PATHEXT;
+        var pathExtArray = splitPath(pathExtEnv);
+        var i;
+
+        // If the extension is already in PATHEXT, just return that.
+        for (i = 0; i < pathExtArray.length; i++) {
+          var ext = pathExtArray[i];
+          if (attempt.slice(-ext.length) === ext && checkPath(attempt)) {
```
Comment (nfischer) [inline src/which.js]: This does a case-sensitive comparison. What we'd really like is a case-insensitive comparison for Windows. Could we perhaps use a regex with the `/i` option?

File: test/which.js [hunk 9]
```diff
@@ -23,8 +23,16 @@ assert.equal(result, null);
...
+
+if (process.platform === 'win32') {
+    // This should be equivalent on Windows
+    var nodeExe = shell.which('node.exe');
+    assert.equal(shell.error(), null);
```
Comment (nfischer) [inline test/which.js]: nit: `assert.ok(!shell.error())`

**Comment by nfischer** [comment]
The nits are all totally optional. They're just for stylistic consistency (which we aren't very good at anyway, as you can probably see by other lines in this test that differ from the style I'm suggesting).

The big issue is the comment about case-insensitive comparison (and also see the typo I pointed out). Once that gets updated, I'll test again and give the thumbs up.

File: src/which.js (hunk 8 above)
Comment (BYK) [inline src/which.js]: I'd prefer running `.toUpperCase()` on the extracted extension.

Comment (nfischer) [inline src/which.js]: I prefer case insensitive comparison. There's no guarantee (as far as I can tell) that the PATHEXT values will **always** be uppercase, and it's a little excessive to call .toUpperCase() on both sides. The regex should be very simple. Whichever is easiest is fine with me though

Comment (isiahmeadows) [inline src/which.js]: Oops...will fix when I get home...

File: test/which.js (hunk 9 above)
Comment (isiahmeadows) [inline test/which.js]: Yeah...I think yours looks better, anyways.

File: test/which.js (hunk 5 above)
Comment (isiahmeadows) [inline test/which.js]: Oops...will fix when I get home...

**Comment by nfischer** [comment]
@isiahmeadows Could you rebase this off master? We now have Windows CI set up with appveyor, and it'd be great to verify this under that (although I'm fairly certain it'll pass after the rebase).

**Comment by isiahmeadows** [comment]
I like that GitHub has this "Update Branch" button where I could do that here. :smile:

**Comment by ariporad** [comment]
What? They added such a button‽‽‽

Thanks! Ari
On Thu, Jan 28, 2016 at 3:38 PM, shelljs/shelljs
<reply@reply.github.com>
wrote:
I like that GitHub has this "Update Branch" button where I could do that here.:smile: [https://assets-cdn.github.com/images/icons/emoji/unicode/1f604.png]

—
Reply to this email directly or view it on GitHub
[https://github.com/shelljs/shelljs/pull/290#issuecomment-176482283] .[https://github.com/notifications/beacon/ABu7pJGYEXZGHWuh2J1oFc_U2wgsCAwAks5pep3-gaJpZM4HFNsu.gif]

**Comment by isiahmeadows** [comment]
That's what I just used. And it doesn't appear to be passing... I'll fix it when I get home. I've been a bit busy lately.

**Comment by isiahmeadows** [comment]
This latest push won't fix the build, but it'll give me a better idea what happened.

**Comment by isiahmeadows** [comment]
@nfischer It's surprising that the AppVeyor build isn't required...I'm guessing that was an oversight? (I just noticed this.)

File: test/which.js (hunk 6 above)
Comment (isiahmeadows) [inline test/which.js]: I removed this because it's redundant after the initial check (if this fails after the last, it's not a bug in this).

**Comment by isiahmeadows** [comment]
I've rebased against `master` to remove all the superflouous commits.

**Comment by isiahmeadows** [comment]
@ariporad I think this is a recent addition, where if you don't have merge conflicts, you can just update your remote branch through GitHub itself instead of locally having to do it. It's just like with merging branches.

**Comment by nfischer** [comment]
@isiahmeadows is checkpath() case insensitive? Could you add a

**Comment by nfischer** [comment]
LGTM (in case lgtm.co missed it)

**Comment by ariporad** [comment]
Thanks @isiahmeadows!

And yeah that was an oversight, At the time we added it, our windows build was broken.

**Comment by isiahmeadows** [comment]
@ariporad Okay.
